share/revng/test/tests/analysis/DetectABI/x86_64/push-pop.S.model.yml

```diff
@@ -5,7 +5,7 @@
 ---
 Functions:
   - OriginalName: push_pop
-    Prototype: "/TypeDefinitions/1-RawFunctionDefinition"
+    Prototype: "/TypeDefinitions[1-RawFunctionDefinition]"
```

Comment:
Did we agree on this syntax?
One of the goals of the path syntax was to let scripting languages rely on `.split("/")`.

Reply:
> Did we agree on this syntax?

The consensus was ["whatever is simpler"](https://mattermost.rev.ng/revng/pl/yqchfc4bkjgz9yx56qcrem6wzc), and I feel like this does make easier to work with. Note that it doesn't have to be `[]`, any distinct separator for "element" as opposed to a "field" works.

1. The logic is already separate, so replacing it with a different separator is a matter of running a regex.
2. Also consider that this makes it easier to deal with/parse, and notice when something is wrong with it.
3. This makes "mask"-style paths make _a lot_ more sense: since now `*` can only ever appear inside `[]`.
4. I don't think it changes the ".split("/")` workflow that much - I'd imagine in most cases you still need to handle containers and fields differently, so this is one more stipulation to make that more explicit: you first run `.split("/")` and then something like `(\w+)(\[([^\]]+)\]|)` for each "leg" of the path to get field name in `match[1]`, confirmation whether this is a container or not in `match[2]` and the key in `match[3]`.

Reply:
Maybe we should not invent a new path syntax. I see there's a [JSON path](https://jsonpath.com/) syntax, of which we could be a strict subset of.

We could have:

1. `$` as a prefix
2. `.` as struct field accessor
3. `[...]` as a dictionary accessor.

Problem is in YAML (/JSON) we have list of objects that we want to index in a custom way, so it's not like the same expression could be used as JSONPath on YAML/JSON.

Maybe there's still some value in this approach in the fact that, if you use our TypeScript/Python wrappers, the path syntax is also valid TypeScript/Python.

Reply:
Sounds good to me (I don't really care which separators we choose, but as long as we have at least 2 distinct ones, I'm happy).

Reply:
> the path syntax is also valid TypeScript/Python

Ah, also, about this, don't forget that we also have polymorphic stuff in: so we'd not be compatible in every case anyway, think
```json
$.TypeDefinitions[0-RawFunctionDefinition].RawFunctionDefinition::StackArgumentsType
```

Reply:
As per the mattermost discussion, this was dropped for now (sadness ensues).